File: Level 7 - System Design & Probabilistic DS/roaring_bitmaps/advanced_problems.java

import java.util.*;

/**
 * Advanced Roaring Bitmap Problems in Java
 * Simulation of a Search Engine Inverted Index using Roaring Bitmaps.
 */
public class advanced_problems {

    static class InvertedIndex {
        Map<String, basic_operations> index = new HashMap<>();

        void indexDoc(int docId, String... tags) {
            for (String tag : tags) {
                index.computeIfAbsent(tag, k -> new basic_operations()).add(docId);
            }
        }

        // Returns documents that match ALL tags (AND operation)
        // Note: Using basic_operations logic but simplified for demonstration
        public List<Integer> search(String... tags) {
            if (tags.length == 0)
                return Collections.emptyList();

            // In a real system, we'd use optimized bitset intersections
            // Here we'll simulate the doc IDs by checking contains()
            List<Integer> results = new ArrayList<>();
            for (int i = 0; i < 10000; i++) {
                boolean match = true;
                for (String tag : tags) {
                    if (!index.containsKey(tag) || !index.get(tag).contains(i)) {
                        match = false;
                        break;
                    }
                }
                if (match)
                    results.add(i);
            }
            return results;
        }
    }

    public static void main(String[] args) {
        InvertedIndex engine = new InvertedIndex();

        System.out.println("Indexing documents with tags...");
        engine.indexDoc(1, "java", "coding", "fast");
        engine.indexDoc(2, "python", "coding", "slow");
        engine.indexDoc(3, "java", "coding", "enterprise");
        engine.indexDoc(4, "cpp", "coding", "fast");

        System.out.println("Search 'java' AND 'coding': " + engine.search("java", "coding"));
        System.out.println("Search 'fast' AND 'coding': " + engine.search("fast", "coding"));
        System.out.println("Search 'enterprise': " + engine.search("enterprise"));
    }
}

File: Level 7 - System Design & Probabilistic DS/roaring_bitmaps/advanced_problems.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>

/**
 * Advanced Roaring Bitmap Problems in C++
 * Simulation of Inverted Index for Tag-based Search.
 */

#include "basic_operations.cpp"

class SearchEngine {
private:
    std::map<std::string, RoaringBitmap> index;

public:
    void addDoc(uint32_t docId, const std::vector<std::string>& tags) {
        for (const auto& tag : tags) {
            index[tag].add(docId);
        }
    }

    std::vector<uint32_t> findByTags(const std::vector<std::string>& tags) {
        std::vector<uint32_t> results;
        if (tags.empty()) return results;

        // Simplified scan (In real Roaring, we'd intersect the Bitmaps)
        for (uint32_t i = 0; i < 1000; ++i) {
            bool all_match = true;
            for (const auto& tag : tags) {
                if (index.find(tag) == index.end() || !index[tag].contains(i)) {
                    all_match = false;
                    break;
                }
            }
            if (all_match) results.push_back(i);
        }
        return results;
    }
};

int main() {
    SearchEngine engine;
    engine.addDoc(100, {"tech", "java"});
    engine.addDoc(200, {"tech", "cpp"});
    engine.addDoc(300, {"finance", "java"});

    std::cout << "Search 'tech' AND 'java':" << std::endl;
    auto res = engine.findByTags({"tech", "java"});
    for (auto id : res) std::cout << "Doc " << id << std::endl;

    return 0;
}
